### DISCUSSION TITLE
NPM 7 is broken

### DISCUSSION THREAD

**Discussion by ghost** [uNetworking/uWebSockets.js#374] [Q&A] [answered]
The geniuses of NPM have broken our installation process with NPM 7,

You can no longer update uWebSockets.js by installing a new version - you have to **uninstall it completely** and then **install it again** for the package to actually "update". Otherwise NPM 7 will say "already up to date" and do nothing.

**Comment by zorro-fr24** [comment]
Speaking of node 15, I don't suppose there is a flying chance that it's worth revisting: https://github.com/uNetworking/uWebSockets.js/issues/341 to see if there is any improvement?
I see there has been a performance related commit to /lib/internal/async_hooks.js

If history is anything to go by though I guess we're looking at another 15% regression

>  async_hooks: improve resource stack performance
> 
> Removes some of the performance overhead that came with
> `executionAsyncResource()` by using the JS resource array
> only as a cache for the values provided by C++. The fact that we now
> use an entry trampoline is used to pass the resource without
> requiring extra C++/JS boundary crossings, and the direct accesses
> to the JS resource array from C++ are removed in all fast paths.
> 
> This particularly improves performance when async hooks are not
> being used.

**Comment by ghost** [comment]
I love your pessimism, but it looks like hell has frozen over!

Node.js 10.22 does 115k, while Node.js 14.5 does 93k - a clear regression. But Node.js 14.14 and Node.js 15.0.0 does about 105k so it's slightly less regressed.

They are still way below where they could have been if they didn't let Kreti & Pleti work on core stuff, but I'll take it!

**Comment by zorro-fr24** [comment, marked as answer]
colour me amazed!
